### PR TITLE
IN: Profile details page #35

### PR DESCRIPTION
**Pull request by luisreyesxv**
### What this PR does (required):
- completed the integration for the profile details page. It is now getting most of it's information from the backend, and is doing a basic creation of a request as well.
- created a basic 404 page with a 404 route. component will push there if a profile isn't found
- changed the Date & Time fields to use material-ui's `<DatePicker / >` & `<TimePicker />` instead of the existing textFields.
-small changes to the request controller so that it doesn't accept userId inside the request body, but instead takes it from the token inside the cookies. (This is safer so we know that a user can't' ever set up a request for another user).
-small change to the "serializers" for Profile and User. wanted to keep it cleaner

### Screenshots / Videos (required):
normal
failed booking
success

### Any information needed to test this feature (required):
- You need to create a profile for a user and have a link handy to get to it.
- you should also run npm install to make sure it catches all the necessary libraries

### Any issues with the current functionality (optional):
- Left a few TODO comments for some missing functionality. Specifically on how to handle available dates. There is a function inside that can be passed into the `DatePicker` that can disable certain dates. The comment explains more for the technique, but we can kind of reverse it and disable every date EXCEPT the ones the profile has available
- Not exactly sure the format needed for start & end dates in request

### CODE REVIEW

File: client/src/components/ProfileDetails/ProfileBookingCard.tsx [hunk 1]
```diff
@@ -1,26 +1,62 @@
...
+  const [dropOffDate, setDropOffDate] = useState<MaterialUiPickersDate>();
+  const [dropOffTime, setDropOffTime] = useState<MaterialUiPickersDate>();
 
-  const changeDropInHandler = (e: React.SyntheticEvent): void => {
-    const target = e.target as HTMLInputElement;
-    setDropIn({ ...dropIn, [target.name]: target.value });
+  const disableDates = (day: MaterialUiPickersDate): boolean => {
+    // if (day) {
+    //   const testDate = new Date('10/7/2021');
+    //   const date = new Date(day?.getDate());
```
Comment (ajaytomgeorge) [inline client/src/components/ProfileDetails/ProfileBookingCard.tsx:34]: For better readability, these commented sections of code can be removed

File: client/src/components/ProfileDetails/ProfileBookingCard.tsx [hunk 2]
```diff
@@ -32,60 +68,59 @@ export default function ProfileBookingCard({ profile }: Props): JSX.Element {
...
-            variant="outlined"
-            id="Drop-off-time"
-            name="time"
-            type="time"
-            InputLabelProps={{ shrink: true }}
-            className={classes.timeFields}
-          />
-        </CardContent>
+        <MuiPickersUtilsProvider utils={DateFnsUtils}>
+          <CardContent>
```
Comment (ajaytomgeorge) [inline client/src/components/ProfileDetails/ProfileBookingCard.tsx:72]: Good work

File: client/src/pages/ProfileDetails/ProfileDetails.tsx [hunk 3]
```diff
@@ -12,29 +13,47 @@ import { Profile } from '../../interface/Profile';
...
+  const history = useHistory();
   const { id } = useParams<{ id: string }>();
   const [profile, setProfile] = useState<Profile | undefined>();
 
   useEffect(() => {
     getProfileById(id).then((res) => {
       if (res.success) {
         setProfile(res.success);
+      } else {
+        history.push('/404');
```
Comment (ajaytomgeorge) [inline client/src/pages/ProfileDetails/ProfileDetails.tsx:25]: Good idea , maybe you want to be more accommodative in naming and info text. Here the assumption indicates that all failures are 404.

File: server/controllers/requests.js [hunk 4]
```diff
@@ -29,41 +31,41 @@ async function updateReqs(req, res, next) {
...
-        const {sitterId, start, end, dogType, specialNotes } = req.body;
-        if (!sitterId|| !start || !end || !req.body.userId) {
-            return res.sendStatus(400)
+        const { sitterId, start, end, dogType, specialNotes } = req.body;
+        if (!sitterId || !start || !end) {
+            return res.status(400).json({ error: "Please make sure all fields are filled out" })
         }
         await Request.create({
-            user: req.body.userId,
+            user: req.user._id,
```
Comment (ajaytomgeorge) [inline server/controllers/requests.js:54]: Good work in changing  to user._id

**Review by ajaytomgeorge — COMMENTED** [review]
Only minor comments

**Comment by luisreyesxv** [comment]
ready for review

File: client/src/components/ProfileDetails/ProfileBookingCard.tsx [hunk 5]
```diff
@@ -1,26 +1,57 @@
...
   const classes = useStyles();
-  const [dropIn, setDropIn] = useState({ date: '', time: '' });
-  const [dropOff, setDropOff] = useState({ date: '', time: '' });
+  const { updateSnackBarMessage } = useSnackBar();
+  const history = useHistory();
+
+  const { user } = profile;
+  // TODO make the default dropIn & dropOff states equal to the first available dates( or whatever makes sense)
+  const [dropInDate, setDropInDate] = useState<MaterialUiPickersDate>();
+  const [dropInTime, setDropInTime] = useState<MaterialUiPickersDate>();
```
Comment (rajivtitus) [inline client/src/components/ProfileDetails/ProfileBookingCard.tsx:27]: You can store the date and time within a single state :)

File: client/src/components/ProfileDetails/ProfileBookingCard.tsx [hunk 6]
```diff
@@ -32,60 +63,59 @@ export default function ProfileBookingCard({ profile }: Props): JSX.Element {
...
+              minDate={new Date()}
+            />
+            <TimePicker
+              required
+              id="Drop-in-time"
+              autoOk
+              InputLabelProps={{ shrink: true }}
+              inputVariant="outlined"
+              className={classes.timeFields}
+              onChange={setDropInTime}
```
Comment (rajivtitus) [inline client/src/components/ProfileDetails/ProfileBookingCard.tsx:88]: You can use the `dropInDate` state here, please see comment above. Ditto for `dropOffTime`

File: client/src/pages/404/Normal404.tsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import useStyles from './useStyles';
+
+export default function Normal404Page(): JSX.Element {
+  const classes = useStyles();
+  const history = useHistory();
+  return (
+    <Grid container component="main" className={classes.root}>
+      <Grid item xs={12} sm={4} className={classes.content}>
+        <Box display="flex" flexDirection="column" justifyContent="center" alignContent="center">
+          <img src={DogPhoto} className={classes.image} />
```
Comment (rajivtitus) [inline client/src/pages/404/Normal404.tsx:13]: It is a good practice to add 'alt' attributes to HTML image tags

File: client/src/pages/ProfileDetails/ProfileDetails.tsx (hunk 3 above)
Comment (rajivtitus) [reply client/src/pages/ProfileDetails/ProfileDetails.tsx:30]: @luisreyesxv  maybe consider updating the snackbar with the error message here?

**Review by rajivtitus — COMMENTED** [review]
Good work here! Just some minor comments :)

**Pull request merged**